### ISSUE TITLE
Look into horovod

### ISSUE THREAD

**Issue by vvksh** [vvksh/learning_stuff#7]
repo: https://github.com/horovod/horovod
based on: Baidu's tensorflow-allreduce algorithm-> read [Bringing HPC Techniques to Deep Learning](https://andrew.gibiansky.com/blog/machine-learning/baidu-allreduce/).

**Comment by vvksh** [comment]
Just read "Bringing HPC Techniques to Deep Learning"; gives detailed outline of ring-allreduce algorithm.

Summary:
When you are doing distributed training training on multiple GPUs, the communication between GPUs especially the amount of data transferred can become a bottleneck. 
*In the straight-forward mechanism where data is sent and received from a single GPU, that single GPU must receive all parameters from all GPUs, and send all parameters to all GPUs. The more GPUs are in the system, the greater the communication cost.* -> grows linearly with number of GPUs

Instead, if you arrange them in ring, and do two rounds of scatter-reduce, and allgather; each round goes on for N-1 iterations, so there will be total of 2(N-1) iterations; hwoever the total data transferred between GPUs overall remains constant; thats because in each iteration, a GPU sends K/N data (K = size of data) , so total data transferred = ![](http://latex2png.com/pngs/3edb2d7113d8694570111be3145f67c6.png) 



Another cool thing: because the gradients are calculated from final layer to initial layer, GPUs can start sending out the gradients when they become available, so you can interleave communication and computation.
